contracts/src/Summa.sol

```diff
@@ -193,28 +215,64 @@ contract Summa is Ownable {
      * @param timestamp The timestamp at which the CEX took the snapshot of its assets and liabilities
      */
     function submitCommitment(
-        bytes memory proof,
+        bytes memory snarkProof,
+        bytes memory grandsumProof,
+        uint256[] memory totalBalances,
         uint256 timestamp
     ) public onlyOwner {
-        require(proof.length == 5376, "Invalid proof length");
+        // Check input lengths
+        require(totalBalances.length > 0, "Invalid total balances length");
+        require(grandsumProof.length == (totalBalances.length * 0x40), "Invalid grandsum proof length");
+        require(snarkProof.length > grandsumProof.length, "Invalid snark proof length");
+        
         uint[] memory args = new uint[](1);
         args[0] = 1; // Workaround to satisfy the verifier (TODO remove after https://github.com/summa-dev/halo2-solidity-verifier/issues/1 is resolved)
         require(
-            polynomialEncodingVerifier.verifyProof(verifyingKey, proof, args),
-            "Invalid proof"
+            polynomialEncodingVerifier.verifyProof(verifyingKey, snarkProof, args),
+            "Invalid snark proof"
         );
-        // TODO slice the proof to get the ID and balance commitments
-        // require(
-        //     balanceCommitments.length == config.cryptocurrencies.length,
-        //     "Liability commitments and cryptocurrencies number mismatch"
-        // );
+        require(
+            totalBalances.length == config.cryptocurrencyNames.length,
+            "Liability commitments and cryptocurrencies number mismatch"
+        );
+
+        bytes memory slicedSnarkProof = new bytes(grandsumProof.length);
+        for (uint256 i = 0; i < grandsumProof.length; i++) {
+            slicedSnarkProof[i] = snarkProof[i + 64]; // Skip first 64 bytes, it's not for total balance.
+        }
+
+        // Concatenate the grandsumProof with snarkProof with same length
+        bytes memory combinedProofs = new bytes(slicedSnarkProof.length + grandsumProof.length);
+        for (uint256 i = 0; i < grandsumProof.length; i++) {
+            combinedProofs[i] = grandsumProof[i];
+        }
+        for (uint256 i = 0; i < slicedSnarkProof.length; i++) {
+            combinedProofs[i + grandsumProof.length] = slicedSnarkProof[i];
+        }
+
+        require(grandSumVerifier.verifyProof(verifyingKey, combinedProofs, totalBalances), "Invalid grandsum proof");
+
+        // Slice the proof aa length as grandsumProof 
+        commitments[timestamp] = slicedSnarkProof;
+
+        emit LiabilitiesCommitmentSubmitted(timestamp, totalBalances, slicedSnarkProof);
+    }
 
-        commitments[timestamp] = proof;
+    function verifyInclusionProof(
+        bytes memory inclusionProof,
+        uint256[] memory challenges,
+        uint256[] memory values
+    ) public view returns (bool) {
+        require(values.length > 1, "Invalid values length");
```

Comment:
We can check the exact values here because we know the `cryptocurrencyNames` length that was passed in the constructor.

Reply:
Ah yes, It been fixed at [0230050](https://github.com/summa-dev/summa-solvency/commit/0230050f47ecae88fd237c1920481b05954e3ef3)